src/server/src/main/java/io/cassandrareaper/service/SegmentGenerator.java

```diff
@@ -81,76 +87,186 @@ static boolean greaterThanOrEqual(BigInteger big0, BigInteger big1) {
+
+    for (Entry<List<String>, List<RingRange>> tokenRangesByReplica : replicasToRange.entrySet()) {
+      LOG.info("Coalescing segments for nodes {}", tokenRangesByReplica.getKey());
+      if (!tokenRangesForCurrentSegment.isEmpty()) {
+        coalescedRepairSegments.add(
+            Segment.builder().withTokenRanges(tokenRangesForCurrentSegment).build());
+        tokenRangesForCurrentSegment = Lists.newArrayList();
+      }
+
+      for (RingRange tokenRange : tokenRangesByReplica.getValue()) {
+        if (tokenRange.span(rangeSize).add(tokenCount).compareTo(targetSegmentSize) > 0
+            && !tokenRangesForCurrentSegment.isEmpty()) {
+          // enough tokens in that segment
+          LOG.info(
+              "Got enough tokens for one segment ({}) : {}",
+              tokenCount,
+              tokenRangesForCurrentSegment);
+          coalescedRepairSegments.add(
+              Segment.builder().withTokenRanges(tokenRangesForCurrentSegment).build());
+          tokenRangesForCurrentSegment = Lists.newArrayList();
+          tokenCount = BigInteger.ZERO;
+        }
+
+        tokenCount = tokenCount.add(tokenRange.span(rangeSize));
+        tokenRangesForCurrentSegment.add(tokenRange);
+
+      }
     }
-    if (!total.equals(rangeSize) && !incrementalRepair) {
-      throw new ReaperException("Not entire ring would get repaired");
+
+    if (!tokenRangesForCurrentSegment.isEmpty()) {
+      coalescedRepairSegments.add(
+          Segment.builder().withTokenRanges(tokenRangesForCurrentSegment).build());
     }
-    return repairSegments;
+
+    int coalescedRanges =
+        coalescedRepairSegments
+            .stream()
+            .map(segment -> segment.getTokenRanges().size())
+            .reduce((first, second) -> first + second)
+            .orElse(0);
+
+    int totalRanges =
+        replicasToRange
+            .values()
+            .stream()
+            .map(List::size)
+            .reduce((first, second) -> first + second)
+            .orElse(0);
+
+    LOG.debug("Coalesced ranges/Total ranges : {}/{}", coalescedRanges);
+
+    LOG.debug("Total number of ranges : {}", totalRanges);
+
+    // Check that we haven't left any token range outside of the resulting segments
+    Preconditions.checkArgument(
+        coalescedRanges == totalRanges,
+        "Number of coalesced tokens doesn't match with the total number of tokens");
+
+    return coalescedRepairSegments;
+  }
+
+  private BigInteger getTargetSegmentSize(int segmentCount) {
+    return (rangeMax.subtract(rangeMin)).divide(BigInteger.valueOf(segmentCount));
   }
 
   protected boolean inRange(BigInteger token) {
     return !(lowerThan(token, rangeMin) || greaterThan(token, rangeMax));
   }
+
+  private boolean supportsSegmentCoalescing(String cassandraVersion) {
+    return !cassandraVersion.startsWith("1.")
+        && !cassandraVersion.startsWith("2.0")
+        && !cassandraVersion.startsWith("2.1");
+  }
```

Comment:
why is this only Cassandra-3+?

and given there's no reason for 256 vnodes in Cassandra-3+ is coalescing an important feature?

Reply:
The method to trigger subrange repairs evolved in 2.2 to accept multiple token ranges. Before that, it would only accept a single range.

Most clusters are running with 256 vnodes, and it is still the official recommendation. Changing the number of vnodes is a complex operation as it requires to create a new DC and decommission the old one, or re-create the cluster from scratch so I do not expect this to change in a near future. 
Clusters will be upgraded from 2.x to 3.x/4.x with the same number of vnodes, and I'm not sure how many people are aware of the new token allocation algorithm.

So I strongly believe that token coalescing will remove the overhead of vnodes on repair for most of Reaper users for at least a few years.

If you have a large number of vnodes, there is an overhead for each as segments are processed through a scheduled loop in Reaper. The default loop interval was moved from 30s to 10s lately, which in a 6 nodes cluster using 256 vnodes makes each repair last 4 hours approximately, at best (1536 segments * 10s). Such clusters could easily be repaired using 20 segments per node instead (to prevent overstreaming) which would bring down the minimum repair time to 20 minutes (and even less if density is low enough to use a smaller number of segments per node).

So yes, I think it's a pretty big deal :)

Reply:
in the `supportsSegmentCoalescing(..)` method can we add a check on a system variable as well please, so that this feature can be disabled if need be (in existing installations).

something like…
```
private static final boolean COALESCING_ENABLED = !Boolean.getBoolean(SegmentGenerator.class.getName() + ".coalescing.disabled");

…

  private static boolean supportsSegmentCoalescing(String cassandraVersion) {
    return COALESCING_ENABLED 
        && !(cassandraVersion.startsWith("1.") || cassandraVersion.startsWith("2.0") || cassandraVersion.startsWith("2.1"));
  }
```

Reply:
I'm not sure it's super necessary, as disabling it is already possible by setting a number of segments that's at least the number of vnodes.

Sure you want this ?

Reply:
yes please. we learnt this lesson from the Preconditions in segment state.